src/ts-ast.ts

```diff
@@ -29,22 +29,25 @@ export class Document {
   }
 
   /**
-   * Invoke `cb` on all nodes in this document, traversing depth-first.
+   * Iterate over all nodes in the document, depth first. Includes all
+   * recursive ancestors, and the document itself.
    */
-  walk(cb: (node: Node) => void) {
-    this.members.map((m) => m.walk(cb));
-    cb(this);
+  * traverse(): Iterable<Node> {
+    for (const m of this.members) {
+      yield* m.traverse();
+    }
+    yield this;
   }
 
   /**
    * Clean up this AST.
    */
   simplify() {
-    this.walk((node: Node) => {
+    for (const node of this.traverse()) {
       if (node.kind === 'union') {
         node.simplify();
```

Comment:
It appears that this works at present because you set this.members to a new array rather than mutating the existing one. I think we should add a note to the implementation of `simplify` that it should be aware of the class's implementation of `traverse`, and generally do replacement of properties rather than mutation.

Reply:
I agree there's some subtlety to this implementation, not sure I understand the exact concern with regard to replacing rather than updating here though, so not sure what to document exactly. There is a note further down that might be related (look for "Note we are not recursing here").

Reply:
Oh, right, traverse goes depth first. Yeah, this is safe then even with mutation, no comment needed.